include/cetl/any.hpp

```diff
@@ -0,0 +1,345 @@
+        }
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return handler_ != nullptr;
+    }
+
+private:
+    // Type-erased handler.
+    using any_handler = void* (*) (detail::action, const any* /*self*/, any* /*other*/);
+
+    // Small Object Optimization (SOO) handler.
+    template <typename Tp>
+    struct soo_handler
+    {
+        static_assert(sizeof(Tp) <= Footprint, "Enlarge the footprint");
+
+        static void* handle(detail::action action, const any* self, any* other)
+        {
+            switch (action)
+            {
+            case detail::action::Get:
+
+                return get(const_cast<any&>(*self));
+
+            case detail::action::Copy:
+
+                copy(*other, *self);
+                return nullptr;
+
+            case detail::action::Move:
+
+                move(*other, const_cast<any&>(*self));
+                return nullptr;
+
+            case detail::action::Destroy:
+
+                destroy(const_cast<any&>(*self));
+                return nullptr;
+
+            default:
+
+                std::terminate();
+            }
+        }
+
+        template <typename... Args>
+        static Tp& create(any& dest, Args&&... args)
+        {
+            Tp* ret = static_cast<Tp*>(static_cast<void*>(&dest.buffer_));
+            new (ret) Tp(std::forward<Args>(args)...);
+            dest.handler_ = handle;
+            return *ret;
+        }
+
+    private:
+        static void* get(any& self)
+        {
+            // TODO: Add RTTI check here.
+            return static_cast<void*>(&self.buffer_);
+        }
+
+        static void copy(any& self, const any& source)
+        {
+            const Tp* src = static_cast<const Tp*>(static_cast<const void*>(&source.buffer_));
+            create(self, *src);
+        }
+
+        static void move(any& self, any& source)
+        {
+            Tp* src = static_cast<Tp*>(static_cast<void*>(&source.buffer_));
+            create(self, std::move(*src));
+            destroy(source);
+        }
+
+        static void destroy(any& self)
+        {
+            Tp* ptr = static_cast<Tp*>(static_cast<void*>(&self.buffer_));
+            ptr->~Tp();
+            self.handler_ = nullptr;
+        }
+
+    };  // struct soo_handler
+
+    template <typename ValueType, typename Any>
+    friend std::add_pointer_t<ValueType> any_cast(Any* operand) noexcept;
+
+    void* handle(detail::action action, any* other = nullptr) const
+    {
+        return handler_ ? handler_(action, this, other) : nullptr;
+    }
+
+    any_handler handler_ = nullptr;
+
+};  // class any
+
+/// \brief Constructs an any object containing an object of type T, passing the provided arguments to T's constructor.
+///
+/// Equivalent to `cetl::any(cetl::in_place_type<ValueType>, std::forward<Args>(args)...)`.
+///
+template <typename ValueType, typename Any, typename... Args>
+inline Any make_any(Args&&... args)
```

Comment:
Template functions are always `inline` by default.

Nitpick but I think we should use `CETL_NODISCARD` here and in similar cases, where not using the return value is clearly an error. It also helps enforce AUTOSAR A0-1-2.

Reply:
will fix it in this pr...